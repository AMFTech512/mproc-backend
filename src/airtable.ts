import Airtable from "airtable";
import Joi from "joi";

export interface AirtableConfig {
  apiKey: string;
}

export function getAirtableConfig(): AirtableConfig {
  return Joi.attempt(
    {
      apiKey: process.env.AIRTABLE_API_KEY,
    },
    Joi.object({
      apiKey: Joi.string().required(),
    })
  );
}

export function initAirtableClient(config?: AirtableConfig) {
  const _config = config || getAirtableConfig();
  return new Airtable({ apiKey: _config.apiKey });
}
