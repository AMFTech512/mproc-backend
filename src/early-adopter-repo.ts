import { Base } from "airtable";
import { DIContainer } from "./di";

export class EarlyAdopterRepo {
  private _base: Base;

  constructor(container: DIContainer) {
    // TODO: move this into a config?
    this._base = container.airtableClient.base("appwx5TpNTb8PdfYi");
  }

  async insertEarlyAdopter(email: string) {
    // TODO: also move this table name into a config?
    return await this._base("Early adopters").create([{ fields: { email } }]);
  }
}
