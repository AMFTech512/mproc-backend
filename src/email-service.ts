import { Resend } from "resend";
import { DIContainer } from "./di";

const DEFAULT_FROM_EMAIL = "Austin Fay <austin@mproc.io>";
const DEFAULT_AUDIENCE_ID = "cf9fb491-4cf4-48f1-98fa-e3c6a6d847d3";

export interface NewContactOpts {
  email: string;
  firstName?: string;
  lastName?: string;
  audienceId?: string;
  unsubscribed?: false;
}

export class EmailService {
  private _resendClient: Resend;

  constructor(container: DIContainer) {
    this._resendClient = container.resendClient;
  }

  async addContact(opts: NewContactOpts) {
    const res = await this._resendClient.contacts.create({
      email: opts.email,
      firstName: opts.firstName,
      lastName: opts.lastName,
      audienceId: opts.audienceId ?? DEFAULT_AUDIENCE_ID,
      unsubscribed: opts.unsubscribed ?? false,
    });

    if (res.error) {
      throw res.error;
    }
  }

  async sendEmail(
    toEmail: string,
    subject: string,
    html: string,
    fromEmail = DEFAULT_FROM_EMAIL
  ) {
    const res = await this._resendClient.emails.send({
      to: toEmail,
      from: fromEmail,
      subject,
      html,
    });

    if (res.error) {
      throw res.error;
    }
  }
}
