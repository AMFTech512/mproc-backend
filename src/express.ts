import express from "express";
import { DIContainer } from "./di";
import Joi from "joi";
import multer from "multer";
import cors from "cors";
import { handleUpload, handleUserCreate } from "./handlers";
import packageJson from "../package.json";

interface ServerConfig {
  port: number;
}

function getServerConfig(): ServerConfig {
  return Joi.attempt(
    {
      port: process.env.SERVER_PORT,
    },
    Joi.object({
      port: Joi.number().required(),
    })
  );
}

export function initExpressApp(container: DIContainer) {
  const app = express();
  const config = getServerConfig();

  const upload = multer({
    dest: "./uploads",
  });

  app.use(cors());

  app.get("/", (req, res) => {
    res.send(`v${packageJson.version} - OK`);
  });

  app.post("/upload", upload.single("file"), handleUpload(container));

  app.post("/user", express.json(), handleUserCreate(container));

  app.listen(config.port, () => {
    console.log(`Server is running on port ${config.port}`);
  });

  return app;
}
