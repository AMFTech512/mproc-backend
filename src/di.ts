import { Application } from "express";
import PQueue from "p-queue";
import { Client } from "pg";
import { JwtConfig } from "./jwt";
import { UserRepo } from "./user-repo";
import { ApiKeyRepo } from "./api-key-repo";
import { KeyUsageLogRepo } from "./key-usage-repo";
import { WebAuthnConfig } from "./webauthn";
import { WebAuthnChallengeRepo } from "./webauthn-challenge-repo";
import { WebAuthnAuthenticatorRepo } from "./webauthn-authenticator-repo";
import { EmailService } from "./email-service";
import { Resend } from "resend";

export interface DIContainer {
  expressApp: Application;
  postgresClient: Client;
  resendClient: Resend;
  userRepo: UserRepo;
  apiKeyRepo: ApiKeyRepo;
  keyUsageLogRepo: KeyUsageLogRepo;
  emailService: EmailService;
  webAuthnChallengeRepo: WebAuthnChallengeRepo;
  webAuthnAuthenticatorRepo: WebAuthnAuthenticatorRepo;
  imagePQueue: PQueue;
  jwtConfig: JwtConfig;
  webAuthnConfig: WebAuthnConfig;
}

export function createDIContainer(): DIContainer {
  return {} as DIContainer;
}
