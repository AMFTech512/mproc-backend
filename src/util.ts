import Joi from "joi";
import cors from "cors";
import { DIContainer } from "./di";

/**
 * Takes an object and returns a set clause string and
 * an array of ordered values to be used in a sql update query.
 * Returns a tuple of [set clause string, values array, next param index].
 * Use the options.exclude parameter to exclude certain keys from the set clause.
 * @param obj
 * @param options
 * @returns
 */
export function makeSetClause(
  obj: object,
  options?: { exclude?: string[] }
): [string, unknown[], number] {
  const parts: string[] = [];
  const values: unknown[] = [];
  let idx = 1;

  Object.entries(obj).forEach(([key, value]) => {
    if (value !== undefined && !options?.exclude?.includes(key)) {
      parts.push(`${key} = $${idx++}`);
      values.push(value);
    }
  });

  return [parts.join(", "), values, idx];
}

export function generateSecureRandomString(length = 16) {
  return crypto
    .getRandomValues(new Uint8Array(length))
    .reduce((acc, byte) => acc + byte.toString(16).padStart(2, "0"), "");
}

/**
 * A custom validator for Joi that checks if a string is valid JSON.
 * @param value
 * @param helpers
 * @returns
 */
export const validateJson: Joi.CustomValidator<string, object> = (
  value,
  helpers
) => {
  try {
    return JSON.parse(value);
  } catch (e) {
    return helpers.error("Invalid JSON");
  }
};

export const oneMonthFromNow = () =>
  new Date(Date.now() + 1000 * 60 * 60 * 24 * 30);
