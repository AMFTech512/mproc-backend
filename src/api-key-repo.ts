import { Client } from "pg";
import { DIContainer } from "./di";
import { generateSecureRandomString } from "./util";
import { createHash } from "crypto";

export interface ApiKeyRow {
  key_hash: Buffer;
  owner_id: string;
  created_at: Date;
  name: string;
  is_active: boolean;
}

export class ApiKeyRepo {
  private _dbClient: Client;

  constructor(container: DIContainer) {
    this._dbClient = container.postgresClient;
  }

  /**
   * Generates a random key and its hash.
   * @returns A tuple of [key, key hash].
   */
  static generateKeyHashPair(): [string, Buffer] {
    const key = generateSecureRandomString(16);
    const keyHash = ApiKeyRepo.hashKey(key);

    return [key, keyHash];
  }

  static hashKey(key: string): Buffer {
    return createHash("sha256").update(key).digest();
  }

  async insert(row: ApiKeyRow) {
    return await this._dbClient.query(
      "INSERT INTO api_keys (key_hash, owner_id, created_at, name, is_active) VALUES ($1, $2, $3, $4, $5)",
      [row.key_hash, row.owner_id, row.created_at, row.name, row.is_active]
    );
  }

  async getByKeyHash(keyHash: Buffer) {
    const res = await this._dbClient.query(
      "SELECT * FROM api_keys WHERE key_hash = $1",
      [keyHash]
    );
    return res.rows[0] as ApiKeyRow;
  }

  async deleteByKeyHash(keyHash: Buffer) {
    return await this._dbClient.query(
      "DELETE FROM api_keys WHERE key_hash = $1",
      [keyHash]
    );
  }
}
