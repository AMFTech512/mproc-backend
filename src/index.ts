import { initAirtableClient } from "./airtable";
import { ApiKeyRepo } from "./api-key-repo";
import { createDIContainer } from "./di";
import { EarlyAdopterRepo } from "./early-adopter-repo";
import { initExpressApp } from "./express";
import { getJwtConfig } from "./jwt";
import { KeyUsageLogRepo } from "./key-usage-repo";
import { initQueue } from "./p-queue";
import { initPostgresClient } from "./postgres";
import { UserRepo } from "./user-repo";
import { getWebAuthnConfig } from "./webauthn";
import { WebAuthnAuthenticatorRepo } from "./webauthn-authenticator-repo";
import { WebAuthnChallengeRepo } from "./webauthn-challenge-repo";

await bootstrap();

async function bootstrap() {
  console.log("Bootstrapping application...");

  // create the DI container
  console.log("Creating DI container...");
  const container = createDIContainer();

  // initialize the jwt config
  console.log("Initializing jwt config...");
  container.jwtConfig = getJwtConfig();

  // initialize the webauthn config
  console.log("Initializing webauthn config...");
  container.webAuthnConfig = getWebAuthnConfig();

  // initialize the postgres client
  console.log("Initializing postgres client...");
  container.postgresClient = await initPostgresClient();

  // initialize the airtable client
  console.log("Initializing airtable client...");
  container.airtableClient = initAirtableClient();

  // initialize the user repo
  console.log("Initializing user repo...");
  container.userRepo = new UserRepo(container);

  // initialize api key repo
  console.log("Initializing api key repo...");
  container.apiKeyRepo = new ApiKeyRepo(container);

  // initialize the key usage log repo
  console.log("Initializing key usage log repo...");
  container.keyUsageLogRepo = new KeyUsageLogRepo(container);

  // initialize webauthn challenge repo
  console.log("Initializing webauthn challenge repo...");
  container.webAuthnChallengeRepo = new WebAuthnChallengeRepo(container);

  // initialize the webauthn authenticator repo
  console.log("Initializing webauthn authenticator repo...");
  container.webAuthnAuthenticatorRepo = new WebAuthnAuthenticatorRepo(
    container
  );

  // initialize the early adopter repo
  console.log("Initializing early adopter repo...");
  container.earlyAdopterRepo = new EarlyAdopterRepo(container);

  // initialize the image process queue
  console.log("Initializing process queue...");
  container.imagePQueue = initQueue();

  // initialize the express app
  console.log("Initializing express app...");
  container.expressApp = initExpressApp(container);
}
