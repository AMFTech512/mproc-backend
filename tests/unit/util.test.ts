import { describe, expect, test } from "bun:test";
import { makeSetClause } from "../../src/util";

describe("util", () => {
  describe("makeSetClause", () => {
    test("should return a valid set clause", () => {
      const obj = {
        id: "123", // should not be included since it's under the list of excluded keys
        name: "foo", // should be included
        age: 42, // should be included
        email: undefined, // should not be included since it's undefined
        phone: null, // should be included since we're explicitly setting to null
      };

      const [clause, values] = makeSetClause(obj, { exclude: ["id"] });

      expect(clause).toMatchSnapshot();
      expect(values).toMatchSnapshot();
    });
  });
});
