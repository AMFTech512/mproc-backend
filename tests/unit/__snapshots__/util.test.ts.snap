// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`util makeSetClause should return a valid set clause 1`] = `"name = $1, age = $2, phone = $3"`;

exports[`util makeSetClause should return a valid set clause 2`] = `
[
  "foo",
  42,
  null,
]
`;

exports[`util makeInsertClause should return a valid insert clause 1`] = `"(name, age, phone) VALUES ($1, $2, $3)"`;

exports[`util makeInsertClause should return a valid insert clause 2`] = `
[
  "foo",
  42,
  null,
]
`;

exports[`util makeInsertClause should return a valid insert clause 3`] = `4`;
